### ISSUE TITLE
Class are not shown, and precision, recall and mAP is zero during and after training

### ISSUE THREAD

**Issue by scm-setpoint** [ultralytics/yolov5#3916]
I want to deploy YoloV5 in a docker container, and therefore all files in the repo are placed in a folder called 'app'. When I run a training session, this is the result in the console:
<img width="1079" alt="image" src="https://user-images.githubusercontent.com/69341060/124730668-da9b9700-df11-11eb-93fb-49398842da68.png">

When I run a training session on the same dataset with identical parameters without placing the repo within an app folder, this is the result in the console:
<img width="1099" alt="image" src="https://user-images.githubusercontent.com/69341060/124730729-e9824980-df11-11eb-9e60-420a3f28364c.png">

Do you have any idea why the first training session cannot produce either the classes or the training results? The only primary difference between the two is the path to the directory:
yolov5/#files
repoName/app/yolov5/#files

Also, is there any way to completely disable wandb without removing it from the source code? It keeps stopping the training session because it cant find the api key.

**Comment by glenn-jocher** [comment]
@scm-setpoint to disable wandb simply run `wandb disabled` or `pip uninstall wandb`.

👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by scm-setpoint** [comment]
Thank you for the quick answer. However, I do not run the training from command line, I call on the main method in train.py directly through an API. 

As for your response regarding the training results, right now I now the results are bad because I'm testing this with a small dataset to save time. The problem is not the bad results however, but that when the repo is placed within an app folder, there are no results at all. And I know this is not right, because the same dataset gives some (although bad) results when trained on a repo not put inside an app folder. The in-app-repo-training cant even print out the classes I have trained, while the out-of-app-repo-training does. Do you know what causes this?

**Comment by glenn-jocher** [comment]
@scm-setpoint Docker images contain latest version of master, code is identical. See Dockerfile for more info.

https://github.com/ultralytics/yolov5/blob/33202b7f0bbfcc55860e8c0cafbf13f227e77a84/Dockerfile#L1-L21

**Comment by scm-setpoint** [comment]
This is the error I get when it tries to logg with wandb:

<img width="814" alt="image" src="https://user-images.githubusercontent.com/69341060/124742853-6a930e00-df1d-11eb-8c60-63fd332b902b.png">

I tried adding wandb='disabled' in the namespace arguments, but it didnt work.

**Comment by glenn-jocher** [comment]
@scm-setpoint `wandb disabled` is a bash command.

@AyushExel might be able to provide some insight into the error message above.

**Comment by scm-setpoint** [comment]
Well as of now, I have disabled wandb through commenting all usage out in train.py. However, the problem with the lacking results still persist. I can't use the standard docker image as I need api functionality. Thats why I put the repo inside a docker image I have made. The fault may come from wrong imports or path declarations, as this is the only primary change between the two repos.

**Comment by AyushExel** [comment]
@scm-setpoint regarding the wandb error, have you tried logging into your account by running either `wandb login` command or `wandb.login()` api call? That error basically means that you're not logged into your W&B account.

**Comment by scm-setpoint** [comment]
@glenn-jocher I don't have a wandb account, and for this purpose I don't really need it. Thats why I want it disabled. But the whole wandb problem is not that important, the problem now is that all results calculated after the training (precision, recall and mAP) are all zero. They are not zero when the same dataset is trained in the original repo.

**Comment by scm-setpoint** [comment]
@glenn-jocher I think I have narrowed down the search a bit. The lines in the image are from test.py:
<img width="960" alt="image" src="https://user-images.githubusercontent.com/69341060/124762603-5eb24680-df33-11eb-9228-9966a690e0eb.png">
When running training with the in-app-repo, line 244 never gets printed, but when I run training with the not-in-app-repo it does. This is because 'stats[0].any()' returns false in the in-app-repo. 

The three prints in lines 237, 239 and 240 shows that the stat objects are populated similarly, although the values themselves differ a bit.

**Comment by glenn-jocher** [comment]
@scm-setpoint per class statistics only show if there are any nonzero values. If mAP is zero (mAP for all classes is always shown), it means all classes have zero mAP, and thus there is no point in viewing this list of zeros.

**Comment by scm-setpoint** [comment]
@glenn-jocher I know that is what the primary function of this is, the problem is only that the mAP should not be zero, as the same dataset and same repo (just not packed inside a folder structure) consistently gives results where the mAP is non-zero. In other words, something wrong is happening with the training process when the yolov5 repo is put inside a folder structure. So far I have tracked it down to this: This line in test.py initializes the object 'current':
<img width="900" alt="image" src="https://user-images.githubusercontent.com/69341060/124771780-b5bc1980-df3b-11eb-8678-6a8d356f24d8.png">
This is then appended with other values to 'stats':
<img width="741" alt="image" src="https://user-images.githubusercontent.com/69341060/124771881-ce2c3400-df3b-11eb-9db3-75e43550a44b.png">
In my little debug print in the image above, the in-folder-repo never prints a positive, while the not-in-folder-repo consistently prints about 5-10 prints.

**Comment by glenn-jocher** [comment]
@scm-setpoint all official environments operate correctly including Docker image.

We are training YOLOv5 models right now in Docker images on AWS EC2 instances which exceed the current public model mAPs in the README.

**Comment by scm-setpoint** [comment]
I created a new repo without packing the yolov5 files inside any folders, and now it works.
